### ISSUE TITLE
Replace naive loops with jax.lax.scan

### ISSUE THREAD

**Issue by fattorib** [fattorib/pmap-transformer#13]
Using ```jax.lax.scan``` results in much quicker compile times. Can be applied to:

- Model forward pass 
- gradient accumulation in train_step

Code References: [here](https://github.com/kingoflolz/mesh-transformer-jax/blob/0f9b555fc1e7f8a94e77347b46dedaae5045811d/mesh_transformer/transformer_shard.py#L491) and [here](https://github.com/kingoflolz/mesh-transformer-jax/blob/0f9b555fc1e7f8a94e77347b46dedaae5045811d/mesh_transformer/transformer_shard.py#L134)
